framework/src/userobjects/PropertyReadFile.C

```diff
@@ -84,13 +96,32 @@ PropertyReadFile::PropertyReadFile(const InputParameters & parameters)
   _bounding_box = MooseUtils::buildBoundingBox(mesh_min, mesh_max);
 
   readData();
+  _current_file_index++;
+  if (_prop_file_names.size() > 1 && _app.isRestarting())
+    mooseError("Multiple files not supported with restart/recover");
+}
+
+void
+PropertyReadFile::initialize()
+{
+  // Set then read new file, only if we have not reached the last file
+  if (_current_file_index < _prop_file_names.size())
+  {
+    _reader.setFileName(_prop_file_names[_current_file_index]);
+    readData();
+    _current_file_index++;
```

Comment:
For the very first `initialize` you shouldn't read the data. Or alternatively you shouldn't read the data in the constructor. I tested this on your `piecewise_constant_elem_multiple.i` input after applying this diff
```
diff --git a/framework/src/auxkernels/FunctionAux.C b/framework/src/auxkernels/FunctionAux.C
index 9a5788f0ac..650ba3bd7c 100644
--- a/framework/src/auxkernels/FunctionAux.C
+++ b/framework/src/auxkernels/FunctionAux.C
@@ -9,6 +9,7 @@
 
 #include "FunctionAux.h"
 #include "Function.h"
+#include "PropertyReadFile.h"
 
 registerMooseObject("MooseApp", FunctionAux);
 
@@ -25,6 +26,7 @@ FunctionAux::validParams()
 FunctionAux::FunctionAux(const InputParameters & parameters)
   : AuxKernel(parameters), _func(getFunction("function"))
 {
+  getUserObjectByName<PropertyReadFile>("reader_element");
 }
 
 Real
diff --git a/test/tests/functions/piecewise_constant_from_csv/piecewise_constant_elem_multiple.i b/test/tests/functions/piecewise_constant_from_csv/piecewise_constant_elem_multiple.i
index 20f22fb908..b5b17bd41e 100644
--- a/test/tests/functions/piecewise_constant_from_csv/piecewise_constant_elem_multiple.i
+++ b/test/tests/functions/piecewise_constant_from_csv/piecewise_constant_elem_multiple.i
@@ -51,6 +51,7 @@
     type = FunctionAux
     variable = 'u'
     function = 'element'
+    execute_on = 'timestep_end'
   []
 []
```
and you get an exodiff. That's because if you apply the true dependency which is that the aux kernel evaluation implicitly depends on the user object execution, then you get a second read of your data before you've evaluated the aux kernel a single time.

Reply:
I like reading in the constructor because it get us access to the data really early. 
I ll add some boolean for initalize

Reply:
well it s not just that I like it, it s relied upon by a lot of the tests.
I ll find something

Reply:
Skipping execution on `EXEC_INITIAL` does not solve the problem; in fact you remove `EXEC_INITIAL` as an available execution flag for this class. My patch still leads to an exodiff even with your newest commit. You will need something like an `_initialize_called` boolean

Reply:
Umm. I tried that first and hit issues.

I ll try figuring out those issues then

Reply:
should be fixed now with force_preaux and no first initialization.

This boolean might not be ideal if the user truly expects a new file read on each execution. I m adding something to the docs